### ISSUE TITLE
Can not deserialize unwrapped list when `@JacksonXmlProperty` localName matches `@JacksonXmlRootElement` localName

### ISSUE THREAD

**Issue by spoon16** [FasterXML/jackson-dataformat-xml#86]
Please refer to http://stackoverflow.com/questions/20673153/using-jackson-to-deserialize-a-collection-of-children-with-the-same-type-as-the

I can not figure out how to deserialize a document like this:

```
<test id="0">
    <test id="0.1">
        <test id="0.1.1" />
    </test>
    <test id="0.2" />
    <test id="0.3">
        <test id="0.3.1" />
    </test>
</test>
```

Where each `test` element is deserialized into a simple POJO.

Here is the code I am dealing with.

```
public class UnitTest {

  @Test
  public void test() throws Exception {
    final ObjectMapper mapper = new XmlMapper();

    final XmlTest before =
        new XmlTest("0", Lists.newArrayList(new XmlTest("0.1", null),
            new XmlTest("0.2", Lists.newArrayList(new XmlTest("0.2.1", null)))));
    final String xml = mapper.writeValueAsString(before);
    System.out.println(xml);
    final XmlTest after = mapper.readValue(xml, XmlTest.class);
  }

  @JacksonXmlRootElement(localName = "test")
  public static class XmlTest {
    @JacksonXmlProperty(localName = "id", isAttribute = true)
    private String id;
    @JacksonXmlElementWrapper(useWrapping = false)
    @JacksonXmlProperty(localName = "test")
    private List<XmlTest> children;

    public XmlTest() {}

    public XmlTest(final String id, final List<XmlTest> children) {
      this.id = id;
      this.children = children;
    }
  }

}
```

My imported dependencies:

```
    compile group: 'com.fasterxml.jackson.dataformat', name: 'jackson-dataformat-xml', version: jacksonVersion
    compile group: 'com.fasterxml.jackson.jaxrs', name: 'jackson-jaxrs-xml-provider', version: jacksonVersion

    jacksonVersion=2.2.+
```

**Comment by JWGmeligMeyling** [comment]
Can confirm issue, any idea how this could be solved?

**Comment by cowtowncoder** [comment]
Unfortunately this test still fails, despite fixes that seemed like they would be relevant (esp #177). Still hope to figure it out eventually.

**Comment by krzysztofpcy** [comment]
Hey, any update on the issue? Any known workaround?

**Comment by Pytry** [comment]
I'm facing this same issue in version 2.10.0. Seems like it is trying to set `private List<XmlTest> children` to a single `XmlTest` instance instead of a `List<XmlTest>`.

**Comment by Pytry** [comment]
I did some experimenting, and created some more examples that demonstrate this behavior.

https://github.com/Pytry/can-not-deserialize-unwrapped-list-when-jacksonxmlproperty-localname-matches-j

This does NOT assert that the described and tested behaviour is correct or a bug.

1. If the localname of the root element in the actual XML is the same as any of it's children, 
then jackson cannot deserialize, regardless of how the class is annotated.

1. If the annotated class is given the same localname as any of it's fields, 
jackson will still deserialize an XML who's actual root element name does not 
match the annotated class.

1. If the XML documents main root element is different than it's children, 
then it's children may have children with a localname that is the same as their parent 
so long as the parent is not the root of the entire document.

**Comment by cowtowncoder** [comment]
I fixed #393, which is probably related, but unfortunately that did not yet resolve this issue.
Hoping to have a look to see if I can fix it.

**Comment by cowtowncoder** [comment]
Initially was planning to fix for 2.12 only, but as per user request, backport in 2.11 for 2.11.1.
